### PR TITLE
Integrate transaction pool to the proposal logic

### PR DESCRIPTION
**Pull request by rphmeier**
On top of #77 

...blocks are currently limited by a size limit (on the transactions within), but it's not a strict requisite for block validity. The BFT subsystem will guard against any blocks with larger sizes getting through. We can make this adaptive.

### CODE REVIEW

File: polkadot/api/src/lib.rs [hunk 1]
```diff
@@ -93,40 +94,64 @@ pub trait BlockBuilder: Sized {
 	fn bake(self) -> Block;
 }
 
+/// A checked block identifier.
+pub trait CheckedBlockId: Clone {
+	/// Yield the underlying block ID.
+	fn block_id(&self) -> &BlockId;
```
Comment (gavofyork) [inline polkadot/api/src/lib.rs]: worth making this implement `AsRef`?

File: polkadot/api/src/lib.rs [hunk 2]
```diff
@@ -141,33 +166,44 @@ macro_rules! with_runtime {
...
+	fn evaluate_block(&self, at: &CheckedId, block: Block) -> Result<()> {
 		with_runtime!(self, at, || ::runtime::system::internal::execute_block(block))
 	}
 
-	fn build_block(&self, parent: &BlockId, timestamp: Timestamp) -> Result<Self::BlockBuilder> {
-		if self.code_at(parent)? != LocalDispatch::native_equivalent() {
-			bail!(ErrorKind::UnknownRuntime);
-		}
+	fn nonce(&self, at: &Self::CheckedBlockId, account: AccountId) -> Result<TxOrder> {
+		with_runtime!(self, at, || ::runtime::system::nonce(account))
```
Comment (gavofyork) [inline polkadot/api/src/lib.rs]: what ever happened to the plan to allow typed dispatching into wasm? poc-2?

File: polkadot/runtime/src/runtime/system.rs [hunk 3]
```diff
@@ -17,19 +17,20 @@
...
+	UncheckedTransaction, Function, InherentFunction, Log
+};
 
-const NONCE_OF: &[u8] = b"sys:non:";
-const BLOCK_HASH_AT: &[u8] = b"sys:old:";
-const TEMP_TRANSACTION_NUMBER: &[u8] = b"temp:txcount:";
+use runtime_io::{print, storage_root, enumerated_trie_root};
+use runtime_support::{Hashable, storage};
+use runtime::{staking, session};
+use storage_keys::{NONCE_OF, BLOCK_HASH_AT, TEMP_TRANSACTION_NUMBER};
```
Comment (gavofyork) [inline polkadot/runtime/src/runtime/system.rs]: as i mentioned before, i'd prefer to keep these defined in their home modules

File: polkadot/transaction-pool/src/lib.rs [hunk 4]
```diff
@@ -0,0 +1,302 @@
...
+	}
+}
+
+/// Scoring implementation for polkadot transactions.
+pub struct Scoring;
+
+impl transaction_pool::Scoring<VerifiedTransaction> for Scoring {
+    type Score = u64;
+
+    fn compare(&self, old: &VerifiedTransaction, other: &VerifiedTransaction) -> Ordering {
```
Comment (gavofyork) [inline polkadot/transaction-pool/src/lib.rs]: spaces!

File: polkadot/api/src/lib.rs (hunk 1 above)
Comment (rphmeier) [reply polkadot/api/src/lib.rs]: potentially, but it's more overhead to implement the trait for any types

File: polkadot/api/src/lib.rs (hunk 2 above)
Comment (rphmeier) [reply polkadot/api/src/lib.rs]: filed an issue: https://github.com/paritytech/polkadot/issues/83

File: polkadot/consensus/src/lib.rs [hunk 5]
```diff
@@ -65,6 +67,9 @@ pub use self::error::{ErrorKind, Error};
 
 mod error;
 
+// block size limit.
+const MAX_TRANSACTIONS_SIZE: usize = 4 * 1024 * 1024;
```
Comment (arkpar) [inline polkadot/consensus/src/lib.rs]: Shouldn't the block size limit be defined by the runtime?

File: polkadot/transaction-pool/src/lib.rs [hunk 6]
```diff
@@ -0,0 +1,302 @@
...
+impl transaction_pool::Scoring<VerifiedTransaction> for Scoring {
+	type Score = u64;
+
+	fn compare(&self, old: &VerifiedTransaction, other: &VerifiedTransaction) -> Ordering {
+		old.inner.transaction.nonce.cmp(&other.inner.transaction.nonce)
+	}
+
+	fn choose(&self, old: &VerifiedTransaction, new: &VerifiedTransaction) -> Choice {
+		if old.inner.transaction.nonce == new.inner.transaction.nonce {
+			Choice::RejectNew // no fees to determine which is better
```
Comment (arkpar) [inline polkadot/transaction-pool/src/lib.rs]: Should probably reject old instead to allow replacing transactions

File: polkadot/consensus/src/lib.rs (hunk 5 above)
Comment (rphmeier) [reply polkadot/consensus/src/lib.rs]: we can move it in there eventually, but I don't see much difference between that and having it negotiated off-chain somehow.

Comment (gavofyork) [reply polkadot/consensus/src/lib.rs]: yeah - should definitely be a chain param defined by the runtime. i guess that's one more thing to go into the runtime/native API.

**Pull request merged**